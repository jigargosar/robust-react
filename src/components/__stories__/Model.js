import {action} from '@storybook/addon-actions'
import {storiesOf} from '@storybook/react'
import {centerPaper} from '../../centerPaper'
import {h} from '../../hyper-script'
import {createFakeModels} from '../__fixtures__/Model.fixture'
import {ModelList} from '../Model'

const story = storiesOf('Components | ModelList', module).addDecorator(
  centerPaper,
)

story.add('basic', () =>
  h(ModelList, {
    models: createFakeModels(),
    onClick: model => event => action('onClick')(model, event.type),
  }),
)
story.add('with primaryRenderer', () =>
  h(ModelList, {
    models: createFakeModels(),
    primaryRenderer: model => `${model.name} (id:${model.id})`,
    onClick: model => event => action('onClick')(model, event.type),
  }),
)
