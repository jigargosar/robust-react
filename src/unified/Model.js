import {ListItemText, MenuItem} from 'material-ui'
import {storiesOf} from '../facade'
import {h} from '../hyper-script'
import {centerPaper} from './story-decorators/centerPaper'
// import 'dom-testing-library/extend-expect'
// import {render} from 'react-testing-library'
// import {describe, expect, it, specs, storiesOf} from '../../../../facade'

const ModelListItem = ({model}) => {
  return h(MenuItem, [h(ListItemText, model.name)])
}

// const createFakeModels = () => {
//   const chance = Chance(11)
//   return R.times(
//     (idx: number) => ({id: idx, name: chance.country({full: true})}),
//     3,
//   )
// }

storiesOf('Unified | Model', module)
  .addDecorator(centerPaper)
  .add('ListItem', () => {
    return h(ModelListItem, {model: {id: 0, name: 'Foo bar'}})
  })
