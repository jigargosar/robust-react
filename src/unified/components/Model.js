import Chance from 'chance'
import {List, ListItemText, MenuItem} from 'material-ui'
import {head, map, propOr, times} from 'ramda'
import {storiesOf} from '../../facade'
import {h} from '../../hyper-script'
import {centerPaper} from './story-decorators/centerPaper'
// import 'dom-testing-library/extend-expect'
// import {render} from 'react-testing-library'
// import {describe, expect, it, specs, storiesOf} from '../../../../facade'

const story = storiesOf('Unified | Model', module).addDecorator(centerPaper)

const createFakes = () => {
  const chance = Chance(11)
  return times(idx => ({id: idx, name: chance.country({full: true})}), 3)
}

const ModelListItem = ({
  model,
  primary = propOr('<primary or name not found>', 'name'),
  onClick = () => () => {},
}) => h(MenuItem, {onClick: onClick(model)}, [h(ListItemText, primary(model))])

story.add('ListItem', () => h(ModelListItem, {model: head(createFakes())}))

const ModelList = ({models, ...other}) =>
  h(List, map(model => h(ModelListItem, {model, ...other}), models))

story.add('List', () => h(ModelList, {models: createFakes()}))

export {ModelList}
