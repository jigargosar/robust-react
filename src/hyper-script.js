import h from 'react-hyperscript'
import hh from 'hyperscript-helpers'

export const {div} = hh(h)


export {h}
