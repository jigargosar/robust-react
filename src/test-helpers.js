import {render} from 'react-testing-library'

export const getRendered = component => render(component).container.firstChild
