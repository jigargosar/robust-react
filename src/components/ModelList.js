import {List, ListItemText, MenuItem} from 'material-ui'
import {map, propOr} from 'ramda'
import {h} from '../hyper-script'

const ModelListItem = ({
  model,
  primaryRenderer = propOr('<primaryRenderer or name not found>', 'name'),
  onClick = () => () => {},
}) =>
  h(MenuItem, {onClick: onClick(model)}, [
    h(ListItemText, {primary: primaryRenderer(model)}),
  ])

const ModelList = ({models, ...other}) =>
  h(List, map(model => h(ModelListItem, {model, ...other}), models))

export {ModelList}
