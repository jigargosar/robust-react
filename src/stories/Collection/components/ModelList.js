// @flow

import {List, ListItemText, MenuItem} from 'material-ui'
import {h} from '../../../hyper-script'

type Model = {name: string}

export type ModelListItemProps = {
  model: Model,
}

const ModelListItem = (props: ModelListItemProps) => {
  return h(MenuItem, [h(ListItemText, props.model.name)])
}

export type ModelListProps = {
  models: Array<Model>,
}

const ModelList = ({models}: ModelListProps) => {
  return h(
    List,
    models.map((model, idx) => h(ModelListItem, {key: idx, model})),
  )
}

export {ModelList, ModelListItem}
