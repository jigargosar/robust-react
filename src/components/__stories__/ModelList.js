import {action} from '@storybook/addon-actions'
import {storiesOf} from '@storybook/react'
import {centerDecorator} from '../../storybook-helpers'
import {h} from '../../hyper-script-utils'
import {createFakeModels} from '../__fixtures__/Model.fixture'
import {ModelList} from '../ModelList'

const story = storiesOf('Components|Model.List', module).addDecorator(
  centerDecorator,
)

story.add('basic', () =>
  h(ModelList, {
    models: createFakeModels(),
    onClick: model => event => action('onClick')(model, event.type),
  }),
)
story.add('with primaryRenderer', () =>
  h(ModelList, {
    models: createFakeModels(),
    primaryRenderer: model => `${model.name} (id:${model.id})`,
    onClick: model => event => action('onClick')(model, event.type),
  }),
)
