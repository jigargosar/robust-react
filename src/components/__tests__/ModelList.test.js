import 'dom-testing-library/extend-expect'
import {head} from 'ramda'

import {render, Simulate} from 'react-testing-library'
import {h} from '../../hyper-script'
import {getRendered} from '../../test-helpers'
import {createFakeModels} from '../__fixtures__/Model.fixture'
import {ModelList} from '../ModelList'

describe('Components', () => {
  describe('ModelList', () => {
    it('should render multiple items without crashing', () => {
      expect(
        getRendered(h(ModelList, {models: createFakeModels()})),
      ).toMatchSnapshot()
    })

    it('should render single items without crashing', () => {
      expect(
        getRendered(h(ModelList, {models: head(createFakeModels())})),
      ).toMatchSnapshot()
    })

    it('should render empty list', () => {
      expect(getRendered(h(ModelList, {models: []}))).toMatchSnapshot()
    })

    it('should call onClick handler with model and event', () => {
      const models = createFakeModels()
      const modelToClick = models[0]

      const {getByText} = render(
        h(ModelList, {
          models,
          onClick: m => e => {
            expect(m).toBe(modelToClick)
            expect(e.type).toEqual('click')
          },
        }),
      )

      Simulate.click(getByText(modelToClick.name))
      expect.assertions(2)
    })
  })
})
