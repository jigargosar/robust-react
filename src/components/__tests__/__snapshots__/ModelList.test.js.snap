// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components ModelList should render empty list 1`] = `
<ul
  class="MuiList-root MuiList-padding"
/>
`;

exports[`Components ModelList should render multiple items without crashing 1`] = `
<ul
  class="MuiList-root MuiList-padding"
>
  <li
    class="MuiButtonBase-root MuiListItem-root MuiListItem-default MuiListItem-gutters MuiListItem-button MuiMenuItem-root"
    role="menuitem"
    tabindex="-1"
  >
    <div
      class="MuiListItemText-root"
    >
      <h3
        class="MuiTypography-root MuiTypography-subheading MuiListItemText-primary"
      >
        Bahamas
      </h3>
    </div>
    <span
      class="MuiTouchRipple-root"
    />
  </li>
  <li
    class="MuiButtonBase-root MuiListItem-root MuiListItem-default MuiListItem-gutters MuiListItem-button MuiMenuItem-root"
    role="menuitem"
    tabindex="-1"
  >
    <div
      class="MuiListItemText-root"
    >
      <h3
        class="MuiTypography-root MuiTypography-subheading MuiListItemText-primary"
      >
        American Samoa
      </h3>
    </div>
    <span
      class="MuiTouchRipple-root"
    />
  </li>
  <li
    class="MuiButtonBase-root MuiListItem-root MuiListItem-default MuiListItem-gutters MuiListItem-button MuiMenuItem-root"
    role="menuitem"
    tabindex="-1"
  >
    <div
      class="MuiListItemText-root"
    >
      <h3
        class="MuiTypography-root MuiTypography-subheading MuiListItemText-primary"
      >
        Oman
      </h3>
    </div>
    <span
      class="MuiTouchRipple-root"
    />
  </li>
</ul>
`;

exports[`Components ModelList should render single items without crashing 1`] = `
<ul
  class="MuiList-root MuiList-padding"
>
  <li
    class="MuiButtonBase-root MuiListItem-root MuiListItem-default MuiListItem-gutters MuiListItem-button MuiMenuItem-root"
    role="menuitem"
    tabindex="-1"
  >
    <div
      class="MuiListItemText-root"
    >
      <h3
        class="MuiTypography-root MuiTypography-subheading MuiListItemText-primary"
      >
        Bahamas
      </h3>
    </div>
    <span
      class="MuiTouchRipple-root"
    />
  </li>
</ul>
`;
